docs/lesson-plans/week-5.md

```diff
@@ -0,0 +1,639 @@
+do that with argument matchers. You can do things like `any()` for anything, `anyInt()` for any integer, 
+`any(MyClass.class)` for that specific class, or even just enter your own number/string for specific entries.
+The result you specify will only be returned/done when that matcher is satisfied. 
+
+>Note: if there's another condition you need satisfied, you can also create your own matcher if the default ones aren't 
+>enough (I've never done that though, the default ones are usually good enough).
+
+Example:
+```java 
+when(myMock.do(anyInt()).thenReturn(5);
+
+myMock.do(5); // returns 5
+myMock.do(-1); // returns 5
+myMock.do(anyInt()); // you can't do this though, it only works in the "when" or "do..." methods
+// This is because you're actually calling the method (or the mocked version), so you need to pass in a real value
+
+when(myMock.other(any(Other.class)).thenReturn(5);
+myMock.other(myOther); // returns 5
+myMock.other(notMyOther); // returns null, we didn't specify what to do here
+```
+
+#### Verification
+
+You can also verify that methods on your mock were called.
+
+Example:
+```java 
+Thing thing = mock(Thing.class);
+thing.callSomeMethod(5);
+
+// This will throw an exception if it wasn't called.
+verify(thing).callSomeMethod(any()); // this also takes argument matchers
+verify(thing).callSomeMethod(6); // this will throw an exception
+```
+
+By default, the above `verify` will only pass if the method was called _exactly_ once. To verify that it wasn't called,
+you can use `verify(thing, never())`. To verify that it was called exactly 5 times, you can use 
+`verify(thing, times(5))`. To verify that it was called at least/most 2 times, you can use 
+`verify(thing, atLeast(2))` or `verify(thing, atMost(2))`. There are other call count matchers available in the Mockito
+documentation.
+
+#### ArgumentCaptors
+
+An `ArgumentCaptor` allows you to capture the argument that was used in a method on a mock. This is usually done
+so that you can perform more exact tests on your mock. 
+
+Example:
+```java 
+// Run your test prerequisites.
+
+// Create your ArgumentCaptor.
+ArgumentCaptor<SomeType> arg = ArgumentCaptor.forClass(SomeType.class);
+
+verify(mock).doSomething(arg.capture());
+
+// Then you can test things about your arg.
+assertEquals("Hey!", arg.getMessage());
+```
+
+#### A Full Mockito Example
+
+Because this topic can be pretty complicated, we'll be providing an extensive example below. Please note, we don't 
+expect you to actually write these tests (you can if you want though). There's not much to test with Mockito on this 
+project (unless you want to mock the 
+database interfaces as an exercise). Writing tests for the router is also a bit much, so Mockito is included in this project more as an
+example of what is available to you if you end up doing more projects in Java. 
+
+Example from `PostsRouterTest`:
```

Comment:
This example seems way too complicated and tied up with the router class that we're not showing. I think it would be a lot clearer if we just created a different method and then tested that to show off mockito concepts. Either way, the reader needs to see the method that's being tested or they'll get confused

Reply:
_N E W E X A M P L E_